### DISCUSSION TITLE
connecting from mobile phone

### DISCUSSION THREAD

**Discussion by Ortapi** [Mair/esp32-course#105] [Course Qeustions]
hi,

i'm working on a little project based on the "website on a chip" from the course.
connected to the ESP32:
ili9341 TFT scree,
ws2812 RGB ledStrip,
LM75,
LED.

i want to broadcast all the data via webPage...

all work fine for now,
getting the temperature, showing it on the screen and all...
the server part goes like that:

a site store on SPIFFS
on startup starting STA service.
connecting to my-esp32.local
input SSID and PASSWORD on a form
then connecting to the local wifi OK.

when i want to connect VIA phone (android) i cant open the page...
the esp32 STA work fine but can get the HTML page...

when the ESP32 already connected to the home wifi it load the page ok but dont get the JS data

**Comment by Mair** [comment]
Try using the IP address. Your phone may not support MDNS

**Reply by Ortapi** [reply]
wow. mindblowing.
my entire project based on getting access to a service HTML page and i see its a global problem with android... :(
that means that i cant even solve it so much.

is there any third party platform you know that can supply the infrastructure and i'll just try to send there via MQTT or any other way?
thanks alot
toda

**Reply by Mair** [reply]
MQTT won't help you load the initial HTML page into the browser. This must come from a web server either on the chip or from the cloud. Out of interest sake, you can load an MQTT client directly into a JavaScript page client on a site but I'm not a fan of this because its not secure and you can't protect it. see [this library](https://github.com/mqttjs/MQTT.js/) on how to do this. Its a much better idea to deal with MQTT communication on a back end server where you can protect it with passwords and certificates for encryption.

The simplest solution is to use a static IP address. Its not elegant but it does work.
The default IP address when the ESP32 is an AP is 192.168.4.1. in other words your site can be reached on http://192.168.4.1 with a phone browser
As a station you can also set a static IP address. Have a look [at this](https://github.com/Mair/esp32-course/blob/master/_13_Internet_Connection/v4.3/13_3_internet_connection/components/wifi/connect.c) at the commented out lines from 166.

If that does not suit you then there are 3 more options
1. Write an app for the phone to use a static AP IP address or BLE,
This is the approach a lot of vendors take but obviously they have the man power, expertise and budget which may not be an option here 

2. Use a generic app for android and IOS that would allow secure communication directly to IOT devices. such as the Nordic BLE app.
This will allow you to create a BLE server on the chip and communicate with it directly. The major disadvantage with this is that you now have a dependency on your customer downloading this app and learning how to use it. A simple app that will allow you to do this generically does not exist (At least not that I have seen) and could be a potential gap in the market 😉

3. Use an online web server server
a. insert the WiFi credentials in some way. (AP with static IP as above or as you already are using a screen so you can prompt the person to insert the WiFi credentials using a keypad or serial link with a PC, etc).
b. Once the device has access to the internet, the phone's browser is redirected to a website you host online.
c. Your back-end of your serve can now communicate via MQTT to the chip. 
The difference here is that chip does not serve the website but is controlled via the websites back end. the advantage is you can use whatever domain name URL you like and you get a lot more control. The disadvantage is much more infrastructure and security required.

Whats best will depend on your use case.

**Reply by Ortapi** [reply]
im actually working on the third option.
for now i've got a website store on a chip that can get WiFi credentials from the user using the ESP as an AP.
already have my mini website store on a server with a deticated domain and can be accessed with any device.
now i need to figure it out how to integrate the MQTT with al of it.

not yet achived to send/rec MQTT with the chip.

i need an MQTT broker to be the middleman...?

i'll see what i can do. 👍
